python_modules/dagster/dagster/_serdes/serdes.py

```diff
@@ -682,9 +682,21 @@ def _pack_value(
             for idx, item in enumerate(cast(list, val))
         ]
     if tval is dict:
+        from dagster._core.definitions.asset_key import AssetKey
+
+        dict_val = cast(dict, val)
+
+        if dict_val and all(type(key) is AssetKey for key in dict_val.keys()):
+            return {
+                "__dict_keyed_by_asset_key_": {
+                    key.to_user_string(): _pack_value(value, whitelist_map, f"{descent_path}.{key}")
+                    for key, value in dict_val.items()
+                }
+            }
+
         return {
             key: _pack_value(value, whitelist_map, f"{descent_path}.{key}")
```

Comment:
apologies for snooping, but the title of this PR was very exciting as this has been a huge pain in the past -- but what happens if `_pack_value()` is called on both the key and the value in all cases (so we don't need to special-case this asset key thing)?

_pack_value() is a no-op on ints, bools, floats, strs, etc., and so this shouldn't have any impact on existing dictionaries (I think), but transparently allows fancier objects to get serde'd properly

Reply:
when you call `_pack_value` on an asset key, it serializes it as a dictionary, which cannot exist as the key of a dictionary